discovery/spec.md

```diff
@@ -778,54 +778,51 @@ upon return of the `202 Accepted` from the original request.
-If an `epoch` is given in the request it MUST be greater than the
-`epoch` of any existing Service of the same `id` or the request MUST fail.
-
-Aside from `id` and `epoch`, unless there is an out of band agreement, all
-mandatory attributes MUST be present in the request, and a Discovery Endpoint
-MUST reject requests that are missing such attributes.
-
-If the Discovery Endpoint is unable to successully add every Service in the
-incoming request then an error MUST be generated and none of the specified
-Services are to be added or updated in the Discovery Endpoint.
-
-The Services in the request MUST be processed in the order in which they are
-given. While any error during the update MUST cause the rollback of any effect
-of the request, the processing of each Service in the request MUST be
-applied as if all previous Services in the request were already completed.
-This means that attribute uniqueness checking, such as on the `name`
-attribute, MUST be done on the new state of all Services as if all previously
-specified Services in the incoming request were already successfully processed.
-
-For example, if a Discovery Endpoint has a Service called `MyService`, it is
-possible to rename that Service to `YourService` and then to add, or update,
-another Service to use `MyService` as its `name`.
-
-Likewise, while it might be less than optimal, it is possible for
-a request to update a Service multiple times if the same `id` is used more
-than once within a request.
+contain all Services in the request or fail. The body of the request message
+MUST contain an array of zero or more Service entries.
+
+The following rules apply to processing the Services specified in the request:
+- any error processing the request MUST result in the entire operation failing
+  and no updates to the Discovery Endpoint are to happen.
+- a Service (as identified by its `id`) MUST NOT appear more than once in the
+  request.
+- any existing Service (as identified by its `id`) MUST be completely
+  replaced by the Service definition in the request, except as noted below.
+  This means that any missing property in the incoming request MUST be
+  treated as a request to delete the property from the existing Service
+  definition.
+- if the incoming Service has an `epoch` value then the request MUST fail if
+  that value is not greater than the current Service's `epoch` value.
+- if the incoming Service does not have an `epoch` value then the Service's
+  `epoch` value MUST be set to a valid value of the Discovery Endpoint's
+  choosing. This means that if there is an existing Service with the same
+  `id` then the value MUST be set to a value greater than the current value.
+- if the incoming Service does not contain an `id` then the Discovery Endpoint
+  MUST assign an appropriate value. The value MUST be globally unique.
+- aside from `id`, `epoch` and `url`, unless there is an out of band agreement,
```

Comment:
The `url` is the URL of the service we're updating so I'm expecting the Discovery Endpoint to set it appropriately. If the client sets it to "ibm.com" I don't think the DE should accept it since we know that's not correct.  And in cases where the DE will pick the ID, there's no way for the client to pass it in. Just seemed easier to have the DE ignore it.

Reply:
I see.  I had not noticed those semantics and filled in with an incorrect understanding.  I understand why you've put this in here (thanks).  I think we need to clarify this in [the definition of the `url` attribute](https://github.com/cloudevents/spec/blob/653aa6424f62c55ae3b00c0a589ed8435381169c/discovery/spec.md#url).

Reply:
To clarify slightly, this optionality of the attribute seems proper but also in contradiction with the attribute definition.  This isn't your fault but it currently says "REQUIRED" as opposed to what `id` says "REQUIRED in responses from the Discovery Endpoint."

Reply:
I decided to attack this from a different angle. I made `id` immutable and `url` read-only. Then I could talk about how these types of attributes can not be modified by clients w/o calling these two out as special in the PUT/POST. See what you think.

Reply:
Interesting.  It does attack the problem from another angle.  I like the new language.  I also think the API specification would benefit from calling out how this behavior shows up in POST and PUT.  I won't push hard on that point.